lib/active_model/serializer/attributes.rb

```diff
@@ -0,0 +1,109 @@
+module ActiveModel
+  class Serializer
+    module Attributes
+      class Attribute
+        delegate :call, to: :reader
+
+        attr_reader :name, :reader
+
+        def initialize(name)
+          @name = name
+          @reader = :no_reader
+        end
+
+        def self.build(name, block)
+          if block
+            AttributeBlock.new(name, block)
+          else
+            AttributeReader.new(name)
+          end
+        end
+      end
+      class AttributeReader < Attribute
+        def initialize(name)
+          super(name)
+          @reader = ->(instance) { instance.read_attribute_for_serialization(name) }
+        end
+      end
+      class AttributeBlock < Attribute
+        def initialize(name, block)
+          super(name)
+          @reader = ->(instance) { instance.instance_eval(&block) }
+        end
+      end
```

Comment:
Sure. I meant defining a ("private") class method as follows:
```ruby
def _attribute_mapping(name, block)
  if block
    ->(instance) { instance.instance_eval(&block) }
  else
    ->(instance) { instance.read_attribute_for_serialization(name) }
  end
end
```
and the `attribute` method would just be
```ruby
+        def attribute(attr, options = {}, &block)
+          key = options.fetch(:key, attr)
+          _attribute_mappings[key] = _attribute_mapping(attr, block)
+        end
```